### PR TITLE
Add jupyter-renderers

### PR DESCRIPTION
**Pull request by gnestor**
Closes https://github.com/jupyterlab/jupyterlab/issues/2587

There's consensus about moving over json-extension and vdom-extension. I also think that geojson-extension should be bundled because ipython now provides `IPython.display.GeoJSON`.

I’m running into some new TS compile errors:

```
../json-extension/src/component.tsx(144,36): error TS2339: Property 'includes' does not exist on type 'string'.
../json-extension/src/component.tsx(172,31): error TS2339: Property 'includes' does not exist on type 'string'.
../json-extension/src/component.tsx(190,15): error TS2339: Property 'includes' does not exist on type 'string'.
../json-extension/src/component.tsx(219,14): error TS2339: Property 'includes' does not exist on type 'string'.
../vdom-extension/src/index.tsx(16,18): error TS7016: Could not find a declaration file for module '@nteract/transform-vdom'. '/Users/grant/Sites/jupyter/jupyterlab/node_modules/@nteract/transform-vdom/lib/index.js' implicitly has an 'any' type.
  Try `npm install @types/@nteract/transform-vdom` if it exists or add a new declaration (.d.ts) file containing `declare module '@nteract/transform-vdom';`
```

### CODE REVIEW

**Comment by blink1073** [comment]
This fixes the vdom error:

```diff
diff --git a/packages/all-packages/src/typings.d.ts b/packages/all-packages/src/typings.d.ts
index 774f20cfe..5a3fe00a7 100644
--- a/packages/all-packages/src/typings.d.ts
+++ b/packages/all-packages/src/typings.d.ts
@@ -6,4 +6,5 @@
 /// <reference path="../../coreutils/typings/url-parse/url-parse.d.ts"/>
 /// <reference path="../../rendermime/typings/ansi_up/ansi_up.d.ts"/>
 /// <reference path="../../terminal/typings/xterm/xterm.d.ts"/>
+/// <reference path="../../vdom/src/transform-vdom.d.ts"/>
 /// <reference path="../../vega2-extension/typings/vega-embed/vega-embed.d.ts"/>
```

**Comment by blink1073** [comment]
Other changes (requires a new `npm install` to pick up symlinks):

```diff
diff --git a/packages/all-packages/tsconfig.json b/packages/all-packages/tsconfig.json
index e175b56f9..5817b75ad 100644
--- a/packages/all-packages/tsconfig.json
+++ b/packages/all-packages/tsconfig.json
@@ -11,6 +11,7 @@
     "lib": [
       "ES5",
       "ES2015.Collection",
+      "ES2015.Core",
       "ES2015.Promise",
       "DOM"
     ],
diff --git a/packages/json-extension/package.json b/packages/json-extension/package.json
index b5012b542..6980347df 100644
--- a/packages/json-extension/package.json
+++ b/packages/json-extension/package.json
@@ -16,7 +16,7 @@
     "mimeExtension": true
   },
   "dependencies": {
-    "@jupyterlab/rendermime-interfaces": "^0.4.0",
+    "@jupyterlab/rendermime-interfaces": "^0.4.1",
     "@phosphor/coreutils": "^1.2.0",
     "@phosphor/widgets": "^1.3.0",
     "react": "~16.0.0",
```

**Comment by blink1073** [comment]
The other two also need their `rendermime-interfaces` deps updated as well.

**Comment by blink1073** [comment]
Snap crackle pop baby:

image

cc @rgbkrk

**Comment by gnestor** [comment]
Boom baby!! 👊

image

File: packages/geojson-extension/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+```bash
+jupyter labextension install @jupyterlab/geojson-extension
+```
+
+## Development
+
+```bash
+# Clone the repo to your local environment
+git clone https://github.com/jupyterlab/jupyter-renderers.git
```
Comment (blink1073) [inline packages/geojson-extension/README.md]: These repo links will need to be updated in all README and package.json files.

File: packages/apputils/package-lock.json [hunk 2]
```diff
@@ -0,0 +1,151 @@
+{
```
Comment (blink1073) [inline packages/apputils/package-lock.json]: this snuck in here

**Comment by blink1073** [comment]
For reference, bundle sizes are:

- 8.41 MB with all three renderers
- 7.99 MB without the geojson renderer
- 7.66 MB prior to this PR

**Comment by gnestor** [comment]
Any thoughts about how to cut down the geojson-extension footprint? I assume leaflet is a relatively large library?

**Comment by blink1073** [comment]
Hmm...

image


image

**Comment by blink1073** [comment]
We could load leaflet asynchronously using `import()` when the class is instantiated.

**Comment by blink1073** [comment]
https://webpack.js.org/guides/code-splitting/#dynamic-imports

**Comment by blink1073** [comment]
Actually, since we're compiling to ES5, it would have to be `require.ensure()`.  We should also name it to distinguish it from the one we're using for CodeMirror [modes](https://github.com/jupyterlab/jupyterlab/blob/79ddfae3606602b20259983d50e2b86ca89d7b9f/packages/codemirror/src/mode.ts#L86) (which should also be using a named `require.ensure`).

**Comment by blink1073** [comment]
For the above errors, we need to check for the existence of the layer before using it in `onAfterAttach/onResize`, because it doesn't exist until the first time we render, which doesn't come until after the document loaded from disk.

**Comment by gnestor** [comment]
Try `git reset`ing to this branch. I pushed the "Fix geojson-extension document rendering" commit that should fix this.

**Comment by rgbkrk** [comment]
Maybe keep the leaflet one out of this PR? Just do JSON + vdom?

**Comment by blink1073** [comment]
Did some experimenting:

-  If we use `require.ensure` we lose the ability to get type information about `leaflet` because `require()` has the signature `(id: string): any;`.
- If we use `import()` we get the type information but it gets transpiled to `Promise.resolve().then(function () { return require(` because we are targeting ES5.  There is an [npm package](https://www.npmjs.com/package/@std/esm) that provides `import()` if TS would let us use it in ES5.
- We could leave it as a synchronous import and split it into a different bundle using an [explicit vendor chunk](https://webpack.js.org/plugins/commons-chunk-plugin/#explicit-vendor-chunk).

My vote is to keep it, as a vendor chunk.

**Comment by blink1073** [comment]
If we agree on the vendor chunk approach, we can do that in a follow up PR that also handles React, CodeMirror, and xterm.

**Comment by blink1073** [comment]
The Travis failure is legit, we need to add `@phosphor/messaging@^1.2.1` to `geojson-extension` `dependencies`.

**Comment by blink1073** [comment]
Here is a version of the geo json renderer the lazily loads the leaflet JS and CSS and supports proper typing.  https://gist.github.com/blink1073/fa5385530e7199b5485c1ee95344406d

**Comment by blink1073** [comment]
Note: we can't use the `import foo = require('foo')` syntax outside of the top level of the module: "error TS1232: An import declaration can only be used in a namespace or module."

**Comment by gnestor** [comment]
Nice, @blink1073!! I assume that I should use this approach with plotly-extension vs. [adding a script tag](https://github.com/jupyterlab/jupyter-renderers/blob/master/packages/plotly-extension/src/index.tsx#L14-L36).

I'm getting the following linter warnings:
```
Cannot find name 'require'.	136:8
Cannot find name 'NodeRequire'.	136:82
```

**Comment by blink1073** [comment]
You need to set `types: ["node"]` in `tsconfig.json`.

**Comment by blink1073** [comment]
Not sure if that will work for plotly because the top level stuff is UMD.

**Comment by blink1073** [comment]
@gnestor, shall we leave GeoJSON out for now and merge the other two?

**Comment by rgbkrk** [comment]
_whispers yes_

I'm hoping to make another `vdom` (for Python) release in the next day or so, I'd love to include mention of the latest JupyterLab support. 😄

**Comment by blink1073** [comment]
We had planned to leave it in master for a bit, so it won't be released this week either way.

**Comment by rgbkrk** [comment]
All good, we'll always have new releases. 😅

**Comment by rgbkrk** [comment]
Woohoo!

**Pull request merged**